File: coding_interview_pattern/two_pointers/python/remove_duplicates.py
#!/usr/bin/python3
"""remove duplicate at most twice in a sorted array
"""
from typing import List

def removeDuplicate(arr: List[int], n: int) -> int:

    if n <= 2: return n

    Len = 0
    arr[Len] = arr[0]
    Len += 1

    for i in range(1, n-1):
        if arr[i-1] != arr[i+1]:
            arr[Len] = arr[i]
            Len += 1
    arr[Len] = arr[n-1]
    Len += 1

    return Len



if __name__ == "__main__":
    arr = [1,1,1,3,3,3,5,5]
    size = len(arr)

    print(arr)

    res = removeDuplicate(arr, size)
    print(res)

    print(arr[:res])

File: coding_interview_pattern/two_pointers/c/remove_duplicates.c
#include <stdio.h>

/**
 * removeDuplicates - remove duplicate at most twice in a sorted array
 *
 * @arr: array of sorted duplicate element
 * @n: size of array
 * Return: new length
 */
int removeDuplicates(int arr[], int n)
{
    if (n <= 2) return n;

    int i, len = 0;
    arr[len++] = arr[0];
    for (i = 1; i < n-1; i++)
    {
       if (arr[i-1] != arr[i+1])
       {
           arr[len++] = arr[i];
       }
    }
    arr[len++] = arr[n-1];

    return len;
}

int main(void)
{
    int arr[] = {5,5,5,6,6,6,8};
    int size = sizeof(arr) / sizeof(arr[0]);

    int newLen = removeDuplicates(arr, size);

    printf("%d\n", newLen);

    for (int i = 0; i < newLen; i++)
        printf("%d ", arr[i]);
    printf("\n");

    return (0);
}
